### PR TITLE
Remove ES2018 transforms for ES2018 target

### LINKED ISSUE #20463: Support for --target ES2018
As TC39 gets ready to finalize ES2018, we should allow pass through for some transformation now only enabled under `--target ESNext`, here is the list:

Disable transforms for:
* [ ] Rest/Spread Properties
* [ ] async generators/`for await ... of`

We also need some new lib definitions:
* [x] Promise finally (https://github.com/Microsoft/TypeScript/issues/20411)

The support for `--target ES2018` was added in https://github.com/Microsoft/TypeScript/pull/20385

### PR DESCRIPTION
**Pull request by saschanaz**
Fixes #20463

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
Seems like there should be a test to ensure that ES2018 actually preserves these features.

**Comment by DanielRosenwasser** [comment]
@rbuckton can you take a look at this PR?

**Comment by saschanaz** [comment]
@DanielRosenwasser Yeah... but where should the test go? Can you point an existing test which can be a guide?

File: src/compiler/types.ts [hunk 1]
```diff
@@ -4902,11 +4902,14 @@ namespace ts {
         // It is a good reminder of how much room we have left
         HasComputedFlags = 1 << 29, // Transform flags have been computed.
 
+        ContainsES2018 = 1 << 30,
```
Comment (rbuckton) [inline src/compiler/types.ts]: We will need to free up some space in transform flags as this is to large of a value. Refer to the comment above on line 4900.

Comment (saschanaz) [reply src/compiler/types.ts]: Maybe it should be done in another PR, before merging this one? Not sure what flags can be removed, though.

Comment (rbuckton) [reply src/compiler/types.ts]: Once #27410 lands, you should be able to choose a bit less than 29.

**Comment by rbuckton** [comment]
@saschanaz: You would likely need to move around some of the tests in `tests/cases/conformance` to align them with the correct ECMAScript edition.

**Comment by saschanaz** [comment]
Couldn't replace `lib: esnext` yet as it depends on #26220.

**Comment by saschanaz** [comment]
So it seems this is blocked again? #28354

**Comment by weswigham** [comment]
That's `TypeFlags` not `TransformFlags`. Different enum.

**Comment by saschanaz** [comment]
Oops!

**Comment by weswigham** [comment]
Ping @rbuckton for review again - this looks pretty done. `esnext` just being optional catch clause right now is pretty great (sad it didn't make it on the 2018 train, would have made this even easier).

**Comment by saschanaz** [comment]
Rebased with git file rename, now it should be easier to review commit-by-commit.

**Comment by fearthecowboy** [comment]
Is this going to make it in 3.3?

**Pull request merged**